### PR TITLE
VDiff2: Support Resuming VDiffs

### PR DESCRIPTION
**Pull request by mattlord**
## Description

This work allows you to explicitly resume an existing `VDiff` workflow using the new `resume <UUID>` `VDiff` action. VDiff will then resume, picking up where it left off and comparing the records where the Primary Key column(s) are greater than the last record processed — with the progress and other status information saved when the run ends. This allows you to:
1. Resume a `VDiff` that may have encountered an ephemeral error (adding an auto-retry on error will come later)
2. Do approximate rolling or differential VDiffs (e.g. done after `MoveTables` finishes the initial copy phase and then again just before `SwitchTraffic`)

In order to better support this for users and testing, we also now show the `CompletedAt` timestamp in the VDiff summary, when it does complete w/o error, and we also show the `RowsCompared` during the run to accurately reflect the work done on resuming.

> ⚠️ We cannot guarantee accurate results for `resume` across major MySQL versions due to implied collation differences. For example: in MySQL 5.7 a table using the `utf8mb4` character set will use the `utf8mb4_general_ci` collation by default whereas in MySQL 8.0 a table using the `utf8mb4` character set will use the `utf8mb4_0900_ai_ci` collation by default. These two collations don't produce the same ordering and thus when resuming from the last PK we processed in the previous `VDiff` run, e.g. using:
> ```
> select c_uuid, created_at, dstuff, dtstuff, dbstuff, cstuff from db_order_test where (c_uuid = 'b169-da8df852-9bad-22000b029685-14e3' and created_at > '2020-05-06 07:22:17') or (c_uuid > 'b169-da8df852-9bad-22000b029685-14e3') order by c_uuid, created_at
> ```
> The results will not be the same. And because we're not diffing the full table, [the reconciliation step we take at the end](https://vitess.io/docs/15.0/reference/vreplication/vdiff/#--max_extra_rows_to_compare) does not help us here as in some cases we will have processed a row on one side only (rather than processing it on both sides but in different orders). The only way to ensure we get the same results on both sides in these cases would be to negotiate the lowest common denominator for the character set between versions — in this case `utf8mb4_general_ci` — and explicitly pass that down on both sides like this:
>```
>select c_uuid, created_at, dstuff, dtstuff, dbstuff, cstuff from db_order_test where (c_uuid = 'b169-da8df852-9bad-22000b029685-14e3' and created_at > '2020-05-06 07:22:17') or (c_uuid > 'b169-da8df852-9bad-22000b029685-14e3') order by c_uuid, created_at collate utf8mb4_general_ci
>```
>_**But that will have a catastrophic impact on performance and overhead**_ as when the collation specified in the query does not match the collation used for the PK index, **_a filesort will be needed to sort the entire resultset before streaming any rows_**. For this reason, we will document the fact that `resume` is not guaranteed to produce accurate results when the source and target tables are not using the same collation.

<details>
<summary>Click here for an example manual test and example results
</summary>

```
## Begin Test
pushd examples/local

./401_teardown.sh; ./101_initial_cluster.sh; ./201_customer_tablets.sh; ./202_move_tables.sh && echo -e "\n---------------------\n\n"

mysql -h 127.0.0.1 -P 15306 -e 'insert into customer values (1, "mlord@planetscale.com"), (2, "mlord@planetscale.com"), (3, "mlord@planetscale.com"), (4, "mlord@planetscale.com")'

mysql -u root --socket=/opt/vtdataroot/vt_0000000200/mysql.sock -e "analyze table vt_customer.customer" >/dev/null

for id in {0..2}; do mysql -u root --socket=/opt/vtdataroot/vt_000000010${id}/mysql.sock -e "set global general_log=ON"; done; for id in {0..2}; do mysql -u root --socket=/opt/vtdataroot/vt_000000020${id}/mysql.sock -e "set global general_log=ON"; done

vtctlclient VDiff -- --v2 --format json --limit 2 customer.commerce2customer create 4c664dc2-eba9-11ec-9ef7-920702940ee0 && sleep 5 && mysql -u root --socket=/opt/vtdataroot/vt_0000000200/mysql.sock --binary-as-hex=false -e "select * from _vt.vdiff_table where table_name='customer'\G" && grep -i select /opt/vtdataroot/*/data/pslord.log | grep -i "order by customer_id" && echo -e "\n---------------------\n\n"

vtctlclient VDiff -- --v2 --format json --limit 3 customer.commerce2customer resume 4c664dc2-eba9-11ec-9ef7-920702940ee0 && sleep 5 && mysql -u root --socket=/opt/vtdataroot/vt_0000000200/mysql.sock --binary-as-hex=false -e "select * from _vt.vdiff_table where table_name='customer'\G" && grep -i select /opt/vtdataroot/*/data/pslord.log | grep -i "order by customer_id" && echo -e "\n---------------------\n\n"

vtctlclient VDiff -- --v2 --format json --limit 4 customer.commerce2customer resume 4c664dc2-eba9-11ec-9ef7-920702940ee0 && sleep 5 && mysql -u root --socket=/opt/vtdataroot/vt_0000000200/mysql.sock --binary-as-hex=false -e "select * from _vt.vdiff_table where table_name='customer'\G" && grep -i select /opt/vtdataroot/*/data/pslord.log | grep -i "order by customer_id" && echo -e "\n---------------------\n\n"

vtctlclient VDiff -- --v2 --format json --limit 4 customer.commerce2customer resume 4c664dc2-eba9-11ec-9ef7-920702940ee0 && sleep 5 && mysql -u root --socket=/opt/vtdataroot/vt_0000000200/mysql.sock --binary-as-hex=false -e "select * from _vt.vdiff_table where table_name='customer'\G" && grep -i select /opt/vtdataroot/*/data/pslord.log | grep -i "order by customer_id" && echo -e "\n---------------------\n\n"

mysql -h 127.0.0.1 -P 15306 -e 'insert into customer values (5, "mlord@planetscale.com"), (6, "mlord@planetscale.com"), (7, "mlord@planetscale.com"), (8, "mlord@planetscale.com")'

mysql -u root --socket=/opt/vtdataroot/vt_0000000200/mysql.sock -e "analyze table vt_customer.customer" >/dev/null

vtctlclient VDiff -- --v2 --format json customer.commerce2customer resume 4c664dc2-eba9-11ec-9ef7-920702940ee0 && sleep 5 && mysql -u root --socket=/opt/vtdataroot/vt_0000000200/mysql.sock --binary-as-hex=false -e "select * from _vt.vdiff_table where table_name='customer'\G" &&  grep -i select /opt/vtdataroot/*/data/pslord.log | grep -i "order by customer_id" && echo -e "\n---------------------\n\n"

vtctlclient VDiff -- --v2 --format json customer.commerce2customer show 4c664dc2-eba9-11ec-9ef7-920702940ee0

mysql -h 127.0.0.1 -P 15306 -e 'insert into customer values (20, "mlord@planetscale.com"), (21, "mlord@planetscale.com")'

mysql -u root --socket=/opt/vtdataroot/vt_0000000200/mysql.sock -e "analyze table vt_customer.customer" >/dev/null

mysql -u root --socket=/opt/vtdataroot/vt_0000000200/mysql.sock -e 'update vt_customer.customer set email="mlord-updated@planetscale.com" where customer_id=20'

vtctlclient VDiff -- --v2 --format json customer.commerce2customer resume 4c664dc2-eba9-11ec-9ef7-920702940ee0 && sleep 5 && mysql -u root --socket=/opt/vtdataroot/vt_0000000200/mysql.sock --binary-as-hex=false -e "select * from _vt.vdiff_table where table_name='customer'\G" && echo -e "\n---------------------\n\n"

vtctlclient VDiff -- --v2 --format json customer.commerce2customer show 4c664dc2-eba9-11ec-9ef7-920702940ee0

popd
## End Test


....


## Start Results
Waiting for workflow to start:

Workflow started successfully with 1 stream(s)

Following vreplication streams are running for workflow customer.commerce2customer:

id=1 on 0/zone1-0000000200: Status: Copying. VStream Lag: 0s.


---------------------


{
	"UUID": "4c664dc2-eba9-11ec-9ef7-920702940ee0"
}
*************************** 1. row ***************************
     vdiff_id: 1
   table_name: customer
        state: completed
       lastpk: fields:{name:"customer_id" type:INT64 table:"customer" org_table:"customer" database:"vt_customer" org_name:"customer_id" column_length:20 charset:63 flags:49667} rows:{lengths:1 values:"1"}
   table_rows: 4
rows_compared: 1
     mismatch: 0
       report: {"TableName": "customer", "MatchingRows": 1, "ProcessedRows": 1, "MismatchedRows": 0, "ExtraRowsSource": 0, "ExtraRowsTarget": 0}
   created_at: 2022-06-26 18:58:26
   updated_at: 2022-06-26 18:58:26
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:26.525291Z	   51 Query	select customer_id, email from customer order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:26.594029Z	   42 Query	select customer_id, email from customer order by customer_id

---------------------


{
	"UUID": "4c664dc2-eba9-11ec-9ef7-920702940ee0"
}
*************************** 1. row ***************************
     vdiff_id: 1
   table_name: customer
        state: completed
       lastpk: fields:{name:"customer_id" type:INT64 table:"customer" org_table:"customer" database:"vt_customer" org_name:"customer_id" column_length:20 charset:63 flags:49667} rows:{lengths:1 values:"3"}
   table_rows: 4
rows_compared: 2
     mismatch: 0
       report: {"TableName": "customer", "MatchingRows": 2, "ProcessedRows": 2, "MismatchedRows": 0, "ExtraRowsSource": 0, "ExtraRowsTarget": 0}
   created_at: 2022-06-26 18:58:26
   updated_at: 2022-06-26 18:58:31
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:26.525291Z	   51 Query	select customer_id, email from customer order by customer_id
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:31.462618Z	   61 Query	select customer_id, email from customer where (customer_id > 1) order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:26.594029Z	   42 Query	select customer_id, email from customer order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:31.528514Z	   54 Query	select customer_id, email from customer where (customer_id > 1) order by customer_id

---------------------


{
	"UUID": "4c664dc2-eba9-11ec-9ef7-920702940ee0"
}
*************************** 1. row ***************************
     vdiff_id: 1
   table_name: customer
        state: completed
       lastpk: fields:{name:"customer_id" type:INT64 table:"customer" org_table:"customer" database:"vt_customer" org_name:"customer_id" column_length:20 charset:63 flags:49667} rows:{lengths:1 values:"4"}
   table_rows: 4
rows_compared: 1
     mismatch: 0
       report: {"TableName": "customer", "MatchingRows": 1, "ProcessedRows": 1, "MismatchedRows": 0, "ExtraRowsSource": 0, "ExtraRowsTarget": 0}
   created_at: 2022-06-26 18:58:26
   updated_at: 2022-06-26 18:58:36
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:26.525291Z	   51 Query	select customer_id, email from customer order by customer_id
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:31.462618Z	   61 Query	select customer_id, email from customer where (customer_id > 1) order by customer_id
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:36.678800Z	   71 Query	select customer_id, email from customer where (customer_id > 3) order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:26.594029Z	   42 Query	select customer_id, email from customer order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:31.528514Z	   54 Query	select customer_id, email from customer where (customer_id > 1) order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:36.726510Z	   66 Query	select customer_id, email from customer where (customer_id > 3) order by customer_id

---------------------


{
	"UUID": "4c664dc2-eba9-11ec-9ef7-920702940ee0"
}
*************************** 1. row ***************************
     vdiff_id: 1
   table_name: customer
        state: completed
       lastpk: fields:{name:"customer_id" type:INT64 table:"customer" org_table:"customer" database:"vt_customer" org_name:"customer_id" column_length:20 charset:63 flags:49667} rows:{lengths:1 values:"4"}
   table_rows: 4
rows_compared: 0
     mismatch: 0
       report: {"TableName": "customer", "MatchingRows": 0, "ProcessedRows": 0, "MismatchedRows": 0, "ExtraRowsSource": 0, "ExtraRowsTarget": 0}
   created_at: 2022-06-26 18:58:26
   updated_at: 2022-06-26 18:58:41
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:26.525291Z	   51 Query	select customer_id, email from customer order by customer_id
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:31.462618Z	   61 Query	select customer_id, email from customer where (customer_id > 1) order by customer_id
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:36.678800Z	   71 Query	select customer_id, email from customer where (customer_id > 3) order by customer_id
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:41.659760Z	   81 Query	select customer_id, email from customer where (customer_id > 4) order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:26.594029Z	   42 Query	select customer_id, email from customer order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:31.528514Z	   54 Query	select customer_id, email from customer where (customer_id > 1) order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:36.726510Z	   66 Query	select customer_id, email from customer where (customer_id > 3) order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:41.754246Z	   78 Query	select customer_id, email from customer where (customer_id > 4) order by customer_id

---------------------


{
	"UUID": "4c664dc2-eba9-11ec-9ef7-920702940ee0"
}
*************************** 1. row ***************************
     vdiff_id: 1
   table_name: customer
        state: completed
       lastpk: fields:{name:"customer_id" type:INT64 table:"customer" org_table:"customer" database:"vt_customer" org_name:"customer_id" column_length:20 charset:63 flags:49667} rows:{lengths:1 values:"8"}
   table_rows: 8
rows_compared: 4
     mismatch: 0
       report: {"TableName": "customer", "MatchingRows": 4, "ProcessedRows": 4, "MismatchedRows": 0, "ExtraRowsSource": 0, "ExtraRowsTarget": 0}
   created_at: 2022-06-26 18:58:26
   updated_at: 2022-06-26 18:58:47
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:26.525291Z	   51 Query	select customer_id, email from customer order by customer_id
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:31.462618Z	   61 Query	select customer_id, email from customer where (customer_id > 1) order by customer_id
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:36.678800Z	   71 Query	select customer_id, email from customer where (customer_id > 3) order by customer_id
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:41.659760Z	   81 Query	select customer_id, email from customer where (customer_id > 4) order by customer_id
/opt/vtdataroot/vt_0000000102/data/pslord.log:2022-06-26T22:58:46.897918Z	   91 Query	select customer_id, email from customer where (customer_id > 4) order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:26.594029Z	   42 Query	select customer_id, email from customer order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:31.528514Z	   54 Query	select customer_id, email from customer where (customer_id > 1) order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:36.726510Z	   66 Query	select customer_id, email from customer where (customer_id > 3) order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:41.754246Z	   78 Query	select customer_id, email from customer where (customer_id > 4) order by customer_id
/opt/vtdataroot/vt_0000000202/data/pslord.log:2022-06-26T22:58:47.007630Z	   89 Query	select customer_id, email from customer where (customer_id > 4) order by customer_id

---------------------


{
	"Workflow": "commerce2customer",
	"Keyspace": "customer",
	"State": "completed",
	"UUID": "4c664dc2-eba9-11ec-9ef7-920702940ee0",
	"RowsCompared": 4,
	"HasMismatch": false,
	"Shards": "0",
	"StartedAt": "2022-06-26 22:58:45",
	"CompletedAt": "2022-06-26 22:58:47"
}
{
	"UUID": "4c664dc2-eba9-11ec-9ef7-920702940ee0"
}
*************************** 1. row ***************************
     vdiff_id: 1
   table_name: customer
        state: completed
       lastpk: fields:{name:"customer_id" type:INT64 table:"customer" org_table:"customer" database:"vt_customer" org_name:"customer_id" column_length:20 charset:63 flags:49667} rows:{lengths:2 values:"21"}
   table_rows: 10
rows_compared: 2
     mismatch: 1
       report: {"TableName": "customer", "MatchingRows": 1, "ProcessedRows": 2, "MismatchedRows": 1, "ExtraRowsSource": 0, "ExtraRowsTarget": 0, "MismatchedRowsSample": [{"Source": {"Row": {"email": "mlord@planetscale.com", "customer_id": "20"}}, "Target": {"Row": {"email": "mlord-updated@planetscale.com", "customer_id": "20"}}}]}
   created_at: 2022-06-26 18:58:26
   updated_at: 2022-06-26 18:58:51

---------------------


{
	"Workflow": "commerce2customer",
	"Keyspace": "customer",
	"State": "completed",
	"UUID": "4c664dc2-eba9-11ec-9ef7-920702940ee0",
	"RowsCompared": 2,
	"HasMismatch": true,
	"Shards": "0",
	"StartedAt": "2022-06-26 22:58:50",
	"CompletedAt": "2022-06-26 22:58:52",
	"TableSummary": {
		"corder": {
			"TableName": "corder",
			"State": "completed",
			"RowsCompared": 0,
			"MatchingRows": 0,
			"MismatchedRows": 0,
			"ExtraRowsSource": 0,
			"ExtraRowsTarget": 0
		},
		"customer": {
			"TableName": "customer",
			"State": "completed",
			"RowsCompared": 2,
			"MatchingRows": 1,
			"MismatchedRows": 1,
			"ExtraRowsSource": 0,
			"ExtraRowsTarget": 0
		}
	},
	"Reports": {
		"corder": {
			"0": {
				"TableName": "corder",
				"ProcessedRows": 0,
				"MatchingRows": 0,
				"MismatchedRows": 0,
				"ExtraRowsSource": 0,
				"ExtraRowsTarget": 0
			}
		},
		"customer": {
			"0": {
				"TableName": "customer",
				"ProcessedRows": 2,
				"MatchingRows": 1,
				"MismatchedRows": 1,
				"ExtraRowsSource": 0,
				"ExtraRowsTarget": 0,
				"MismatchedRowsSample": [
					{
						"Source": {
							"Row": {
								"customer_id": "20",
								"email": "mlord@planetscale.com"
							}
						},
						"Target": {
							"Row": {
								"customer_id": "20",
								"email": "mlord-updated@planetscale.com"
							}
						}
					}
				]
			}
		}
	}
}
## End Results
```

</details>

## Related Issue(s)
  - Fixes: https://github.com/vitessio/vitess/issues/10495

## Checklist

-   [x] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added
-   [x] Documentation: https://github.com/vitessio/website/pull/1060

### CODE REVIEW

File: go/test/endtoend/vreplication/vdiff_helper_test.go [hunk 1]
```diff
@@ -109,30 +124,46 @@ type expectedVDiff2Result struct {
...
 			require.False(t, info.HasMismatch)
-
 		}
+		if strings.Contains(t.Name(), "AcrossDBVersions") {
+			log.Errorf("VDiff resume cannot be guaranteed between major MySQL versions due to implied collation differences, skipping resume test...")
+			return
+		}
+		createCompletedTS, err := time.Parse(tsFormat, info.CompletedAt)
+		require.NoError(t, err)
+
```
Comment (rohit-nayak-ps) [inline go/test/endtoend/vreplication/vdiff_helper_test.go:151]: Will any of the "resumes" in the e2e tests see new data to diff or is the diff fully completed due to the previous "create" action?

**Review by rohit-nayak-ps — COMMENTED** [review]
Overall looks great! Nice to see resume land, which was one of the main reasons for vdiff2 :-)

File: go/test/endtoend/vreplication/vdiff_helper_test.go [hunk 2]
```diff
@@ -109,30 +124,46 @@ type expectedVDiff2Result struct {
...
 	ksWorkflow := fmt.Sprintf("%s.%s", keyspace, workflow)
 	t.Run(fmt.Sprintf("vdiff2 %s", ksWorkflow), func(t *testing.T) {
-		uuid, _ := performVDiff2Action(t, ksWorkflow, "create", "")
-		info := waitForVDiff2ToComplete(t, ksWorkflow, uuid)
+		uuid, _ := performVDiff2Action(t, ksWorkflow, cells, "create", "")
+		info := waitForVDiff2ToComplete(t, ksWorkflow, cells, uuid, time.Time{})
 
 		require.Equal(t, workflow, info.Workflow)
 		require.Equal(t, keyspace, info.Keyspace)
+		// I'm not sure if we always have rows in every table
```
[the remark this answers is not in the document]
Comment (mattlord) [reply go/test/endtoend/vreplication/vdiff_helper_test.go:135]: VDiff is called in far more workflows than I expected. Are you asking if I can go through each table its run on and make sure it has at least 1 row? I can, just want to confirm.

File: go/test/endtoend/vreplication/vdiff_helper_test.go (hunk 1 above)
Comment (mattlord) [reply go/test/endtoend/vreplication/vdiff_helper_test.go:151]: When using `--tablet_types=primary` and the `catchup()` function to ensure we get past the copy phase we don't process any additional rows. The vreplication_* workflows were passing once I added missing catchup calls and added that flag to `VDiff --v2`, but then other tests e.g. some vstream_* tests started failing. Do you want me track all of those down?

File: go/test/endtoend/vreplication/vdiff_helper_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rohit-nayak-ps) [reply go/test/endtoend/vreplication/vdiff_helper_test.go:135]: We do have one specific vdiff2 test, so no need to do that. Was just checking.

File: go/vt/vttablet/tabletmanager/vdiff/schema.go [hunk 3]
```diff
@@ -64,26 +75,33 @@ const (
...
-										from _vt.vdiff v, _vt.vdiff_table vt 
-										where v.id = vt.vdiff_id and v.id = %d`
-	sqlUpdateVDiffState     = "update _vt.vdiff set state = %s where id = %d"
+	sqlVDiffSummary                   = `select vd.state as vdiff_state, vdt.table_name as table_name,
+										vd.vdiff_uuid as 'uuid', vdt.state as table_state, vdt.table_rows as table_rows,
+										vd.started_at as started_at, vdt.rows_compared as rows_compared, vd.completed_at as completed_at,
+										IF(vdt.mismatch = 1, 1, 0) as has_mismatch, vdt.report as report
+										from _vt.vdiff as vd inner join _vt.vdiff_table as vdt on (vd.id = vdt.vdiff_id)
+										where vdt.vdiff_id = %d`
+	// sqlUpdateVDiffState has a penultimate placeholder for any additional columns you want to update, e.g. `, foo = 1`
```
Comment (rohit-nayak-ps) [inline go/vt/vttablet/tabletmanager/vdiff/schema.go:90]: :+1:

File: go/vt/vtctl/vdiff2.go [hunk 4]
```diff
@@ -385,56 +405,90 @@ func buildVDiff2SingleSummary(wr *wrangler.Wrangler, keyspace, workflow, uuid st
...
+					reports[table][shard] = dr
+					tableSummaryMap[table] = ts
+
+					// The global VDiff summary needs to align with the table states across all
+					// shards. It should progress from pending->started->completed with error at any point
+					// being sticky. This largely mirrors the global table summary, with the exception
+					// being the started state, which should also be sticky; i.e. we shouldn't go from
+					// started->pending->started in the global VDiff summary state.
+					if summary.State != ts.State &&
+						(summary.State != vdiff.StartedState && ts.State != vdiff.PendingState) {
```
[the remark this answers is not in the document]
Comment (rohit-nayak-ps) [reply go/vt/vtctl/vdiff2.go:488]: after discussion, current logic is fine.

File: go/vt/vtctl/vdiff2.go [hunk 5]
```diff
@@ -385,56 +405,90 @@ func buildVDiff2SingleSummary(wr *wrangler.Wrangler, keyspace, workflow, uuid st
...
+					// known state with completed.
+					switch sts {
+					case vdiff.CompletedState:
+						if ts.State == vdiff.UnknownState {
+							ts.State = sts
+						}
+					case vdiff.ErrorState:
+						ts.State = sts
+					default:
+						if ts.State != vdiff.ErrorState {
```
[the remark this answers is not in the document]
Comment (rohit-nayak-ps) [reply go/vt/vtctl/vdiff2.go:456]: after discussion, current logic is fine.

File: go/test/endtoend/vreplication/vdiff2_test.go [hunk 6]
```diff
@@ -30,32 +30,40 @@ type testCase struct {
...
+		typ:           "MoveTables",
+		sourceKs:      "product",
+		targetKs:      "customer",
+		sourceShards:  "0",
+		targetShards:  "-80,80-",
+		tabletBaseID:  200,
+		tables:        "customer,Lead,Lead-1",
+		resume:        true,
+		resumeInsert:  `insert into customer(cid, name, typ) values(12345678, 'Testy McTester', 'soho')`,
+		testCLIErrors: true, // test for errors in the simplest test case
```
[the remark this answers is not in the document]
Comment (mattlord) [reply go/test/endtoend/vreplication/vdiff2_test.go:51]: I had initially gone that route too, but then there's no workflows to test against — as `testWorkflow()` does `SwitchWrites` and `Complete` at the end — and the only errors you get are a result of the fact that the workflow does not exist. :-)

File: go/vt/vtctl/vdiff2.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mattlord) [reply go/vt/vtctl/vdiff2.go:456]: No, the table diff as a whole (summary here) has started if it's started on any shard.

File: go/vt/vtctl/vdiff2.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattlord) [reply go/vt/vtctl/vdiff2.go:488]: No, the vdiff as a whole (summary here) has started if we started any table on any shard. Unless I misunderstood the question?

**Review by rohit-nayak-ps — APPROVED** [review]
looks great to me!

**Pull request merged**